Mathlib/Data/ENNReal/Floor.lean

```diff
@@ -0,0 +1,162 @@
+
+end ENat_topology
+
+section ENNReal_floor
+
+namespace ENNReal
+
+/-- The floor function `ℝ≥0∞ → ℕ∞`: the floor of `x` is the supremum of the extended natural
+numbers `n` satisfying `n ≤ x`. -/
+noncomputable def floor (x : ℝ≥0∞) : ℕ∞ := sSup {n : ℕ∞ | n ≤ x}
+
+variable {x y : ℝ≥0∞} {n : ℕ} {m : ℕ∞}
+
+lemma floor_eq_sSup_range_toENNReal_inter_Iic :
+    floor x = sSup (Set.range (fun (m : ℕ∞) ↦ (m : ℝ≥0∞)) ∩ Iic x) := by
+  simp only [floor]
+  rw [toENNReal_mono.map_sSup_of_continuousAt ENat.continuous_toENNReal.continuousAt (by simp)]
+  congr
+  ext m
+  simp only [mem_image, mem_setOf_eq, mem_inter_iff, mem_range, mem_Iic]
+  refine ⟨fun hx ↦ ?_, fun hx ↦ ?_⟩
+  · obtain ⟨n, hn⟩ := hx
+    refine ⟨⟨n, hn.2⟩, hn.2 ▸ hn.1⟩
+  · obtain ⟨n, hn⟩ := hx.1
+    refine ⟨n, ⟨hn ▸ hx.2, hn⟩⟩
+
+lemma floor_mono (h : x ≤ y) : x.floor ≤ y.floor := sSup_le_sSup <| fun _ hx ↦ hx.trans h
+
+/-- The floor function `ℝ≥0∞ → ℝ≥0∞` is increasing. -/
+lemma monotone_floor : Monotone floor := fun _ _ h ↦ floor_mono h
+
+lemma floor_le (x : ℝ≥0∞) : floor x ≤ x := by
+  simpa only [floor_eq_sSup_range_toENNReal_inter_Iic] using sSup_le fun _ h ↦ h.2
+
+lemma le_floor (h : m ≤ x) : m ≤ x.floor := le_sSup <| by simp [h]
+
+lemma floor_le_of_lt_add_one (h : x < m + 1) :
+    x.floor ≤ m := by
+  cases' m with l
+  · exact le_top
+  · apply sSup_le
+    intro n hn
+    have obs := lt_of_le_of_lt hn h
+    cases' n with k
+    · exfalso; apply not_top_lt obs
+    · exact Nat.cast_le.mpr <| Nat.le_of_lt_succ (by exact_mod_cast obs)
+
+@[simp] lemma floor_coe_eq (n : ℕ∞) :
+    floor n = n :=
+  le_antisymm (by exact_mod_cast floor_le n) (le_sSup (by simp))
+
+@[simp] lemma floor_zero_eq : floor 0 = 0 := by exact_mod_cast ENNReal.floor_coe_eq 0
+
+@[simp] lemma floor_one_eq_one : floor 1 = 1 := by exact_mod_cast ENNReal.floor_coe_eq 1
+
+@[simp] lemma floor_top_eq : floor ∞ = ⊤ := floor_coe_eq ⊤
+
+lemma floor_eq_natFloor_toNNReal (x_ne_top : x ≠ ∞) :
+    floor x = Nat.floor x.toNNReal := by
+  simp only [floor]
+  apply le_antisymm
+  · apply sSup_le
+    intro b hb
+    by_contra con
+    have key : ⌊x.toNNReal⌋₊ + 1 ≤ b := add_one_le_of_lt (not_le.mp con)
+    have oops : x.toNNReal < x :=
+      lt_of_lt_of_le (by exact_mod_cast Nat.lt_floor_add_one x.toNNReal)
+        (show ⌊x.toNNReal⌋₊ + 1 ≤ x from le_trans (by exact_mod_cast key) hb)
+    exact (lt_self_iff_false x).mp (by simp [ENNReal.coe_toNNReal x_ne_top] at oops)
+  · apply le_sSup
+    simp only [mem_setOf_eq, toENNReal_coe]
+    apply le_trans _ <| coe_toNNReal_le_self (a := x)
+    exact_mod_cast Nat.floor_le (show 0 ≤ x.toNNReal by simp)
+
+lemma floor_floor (x : ℝ≥0∞) : (x.floor : ℝ≥0∞).floor = x.floor := by simp
+
+lemma floor_lt_top {x : ℝ≥0∞} (x_ne_top : x ≠ ∞) :
+    x.floor < ∞ := by
+  simpa [floor_eq_natFloor_toNNReal x_ne_top] using (natCast_ne_top ⌊x.toNNReal⌋₊).symm.lt_top'
+
+@[simp] lemma floor_add_one {x : ℝ≥0∞} :
+    (x + 1).floor = x.floor + 1 := by
```

Comment:
Can you try following the `Nat.floor` API? This involves proving `floor_add_nat` (maybe renamed to `floor_add_natCast` before `floor_add_one`